### ISSUE TITLE
Data types not preserved in categoricals

### ISSUE THREAD

**Issue by story645** [matplotlib/matplotlib#9350]
@anntzer pointed out on #9340 that the behavior for `plt.plot([23, "23"])` should be two data points with identical labels. Currently it collapses them into one value:
![index](https://user-images.githubusercontent.com/1300499/31408524-e700edda-add6-11e7-9186-38887667da69.png)

**Comment by tacaswell** [comment]
I am not super concerned about this, it is string catagoricals.

**Comment by efiring** [comment]
Which brings us back to @anntzer's earlier suggestion that we should raise an error instead of accepting this nonsensical input.

**Comment by anntzer** [comment]
As an additional point where allowing uncontrolled mixing of strings and numbers can be puzzling:
```
fill_between([0, 1, 2], [2, 3, 2])
```
great, let's plot this in red
```
fill_between([0, 1, 2], [2, 3, 2], "r")
```
(saved on the same figure to save some room)
![figure_1](https://user-images.githubusercontent.com/1322974/31812042-904ee458-b536-11e7-82c2-8e0b8b1b7de5.png)

Ah, that's because the second parameter is "y2" (so I needed to write `color="r"`) and "r" got converted to "0" by the categorical converter... (I would say once an axes has used the scalar (passthrough) converter it should disable the categorical converter, just as the categorical converter should not pass scalars through.)

I knew exactly the issue with categoricals, and that behavior still led to a bit a head scratching...

**Comment by story645** [comment]
I agree that's a bug, but the issue seems to be that the second positional parameter is ambiguous. What I mean is 
```python
plt.fill_between(['a', 'b', 'c'], ['e', 'f', 'g'], 'r')
```
Errors out spectacularly because it tries to do a lookup on 'r' against the existing mapping and doesn't see it as a new value:
```
~/projects/matplotlib/lib/matplotlib/category.py in convert(value, unit, axis)
     47 
     48         if isinstance(value, six.string_types):
---> 49             return vmap[value]
     50 
     51         str_value = shim_array(value)

KeyError: 'r'
```

**Comment by tacaswell** [comment]
Closed by #9783

Still some broken edge cases (which we have tests for now).
